staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/peterbourgon/diskv"
```

Comment:
diskv is thread-safe AFAICT, do you need to be thread-safe too, I would think so.

Reply:
We could implement a lock to ensure the consistency between the wrapped diskv and the denormalized file metadata and current cache size, but I'm not sure this is necessary. diskv concurrency guarantees ensure the atomicity of the key (filename) and file contents being written. This ensures the integrity of a single entry in the diskv cache. It is especially important that the file contents are not corrupted.

The `SizeLimitedDiskCache` wraps diskv and stores some denormalized data about files in the cache directory. This denormailzed data is basically used to keep track of the cache size (as well as the file modification times). So what happens if this denormalized data is inconsistent with the entries in diskv? If we believe the the cache size is larger than it really is, then we could evict some of the older files--overly aggressive cache eviction. If we believe the cache size is smaller than it really is, then we might not evict older files which should be evicted--the max cache size is slightly breached. Neither of these conditions seem too concerning.

Let me know what you think.